synapse/handlers/federation.py

```diff
@@ -171,6 +171,23 @@ def on_receive_pdu(self, origin, pdu, get_missing=True):
                         yield self._get_missing_events_for_pdu(
                             origin, pdu, prevs, min_depth
                         )
+
+                        # Update the set of things we've seen after trying to
+                        # fetch the missing stuff
+                        have_seen = yield self.store.have_events(
+                            [ev for ev, _ in pdu.prev_events]
+                        )
+
+                        seen = set(have_seen.keys())
```

Comment:
`have_seen` has been updated the line above, and `have_seen` needs to be rechecked as we just fetched some missing events. Is that what you mean? Or do you mean we should wrap the call to `have_seen` and `seen` in a function?